src/access/RequiredPrivilege.cpp

```diff
@@ -0,0 +1,214 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "RequiredPrivilege.h"
+
+namespace {
+
+using namespace chip;
+using namespace chip::Access;
+
+// Privilege override entries are stored in a table per operation (read attribute,
+// write attribute, invoke command, read entry). Entries in each table are sorted
+// by their primary compound key of cluster/endpoint/field, and therefore cannot
+// have duplicate privilege for any such combination. Cluster cannot be invalid,
+// so that is used as a sentinel for empty/unused entries, which are kept at the
+// end of the table. Endpoint and field can be invalid, which means wildcard.
+// Sorting the entries in this way allows a cluster to override specific endpoints
+// or fields with privilege, while still using wildcard to override all the
+// unspecified ones.
+struct PrivilegeOverride
+{
+    ClusterId mCluster;
+    EndpointId mEndpoint;
+    FieldId mField;
+    Privilege mPrivilege;
+
+    PrivilegeOverride() : mCluster(kInvalidClusterId), mEndpoint(kInvalidEndpointId), mField(kInvalidFieldId) {}
+
+    PrivilegeOverride(ClusterId cluster, EndpointId endpoint, FieldId field, Privilege privilege) :
+        mCluster(cluster), mEndpoint(endpoint), mField(field), mPrivilege(privilege)
+    {}
+
+    static_assert(sizeof(FieldId) >= sizeof(AttributeId), "FieldId must be able to hold AttributeId");
+    static_assert(sizeof(FieldId) >= sizeof(CommandId), "FieldId must be able to hold CommandId");
+    static_assert(sizeof(FieldId) >= sizeof(EventId), "FieldId must be able to hold EventId");
+};
+
+PrivilegeOverride
+    privilegeOverrideForReadAttribute[CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_READ_ATTRIBUTE] = {
+        PrivilegeOverride(0x0000'001f, kInvalidEndpointId, kInvalidFieldId, Privilege::kAdminister),
```

Comment:
Please use the namespaced cluster ids instead of `0x0000'001f`, so it's clear what overrides are being defined.

I guess that would make it harder to tell whether things are sorted... Perhaps we should put the name in a comment (or put the name in the value and the numeric value in a comment, as preferred).

If we're going to use this for real, not just for the one-entry case, we could also write a constexpr IsSorted function, and then leave things with symbolic values and static_assert that it's sorted.  But for now that seems like overkill, as does the general discussion of sorting.

Reply:
Yes I thought of static asserting that the list is sorted, etc. and those are good things to do, just didn't get to it while making this straw-man PR. If we like this approach, we should certainly add improvements like that.

Constants for cluster IDs... where will those be defined? Access control is down at the system level, and doesn't really care about the actual values of the things it is checking (kinda like neurons don't care what they are thinking...). Maybe I should just hoist this class up into app where all that cluster stuff lives.

Reply:
I moved the class from src/access into src/app, made it use the namespaced generated cluster ID, and removed the requirement for the table to be sorted. It still should have more specific entries before more general entries, I added WARNINGs to that effect, if you can write some constexpr magic to check that, that would be nice, but I don't think it's a super high priority at the moment. I think if we get to the point where this class becomes less temporary and more permanent (not a desirable outcome!) then we'll have to give it a once-over to decide what we need from it.